pkg/logger/ecs_error_log.go

```diff
@@ -0,0 +1,218 @@
+	ErrorFileMaxSize int
+
+	errorChannel chan zapcore.Entry
+	queue        []zapcore.Entry
+	ticker       time.Ticker
+	errorFile    *os.File
+	size         int
+}
+
+func NewECSErrorLogger() *ECSErrorLogger {
+	ecsErrorFilePath := os.Getenv("STATUS_MESSAGE_FILE_PATH")
+	errorLogMaxAge, err := strconv.Atoi(os.Getenv("STATUS_MESSAGE_TTL"))
+	if err != nil {
+		log.Fatalf("failed to get STATUS_MESSAGE_TTL: %v", err)
+	}
+	errorFileSize, err := strconv.Atoi(os.Getenv("STATUS_MESSAGE_MAX_BYTE_LENGTH"))
+	if err != nil {
+		log.Fatalf("failed to get STATUS_MESSAGE_MAX_BYTE_LENGTH: %v", err)
+	}
+	return &ECSErrorLogger{
+		ErrorFilePath:    ecsErrorFilePath,
+		ErrorLogMaxAge:   errorLogMaxAge,
+		ErrorFileMaxSize: errorFileSize,
+		errorChannel:     make(chan zapcore.Entry, 100),
+	}
+}
+
+// Run func help select the different case between either write error log into file or
+// rotate and delete old logs in file based on TTL
+func (l *ECSErrorLogger) Run() {
+	defer l.ticker.Stop()
+	for {
+		select {
+		case newError := <-l.errorChannel:
+			l.Write(newError)
+		case <-l.ticker.C:
+			l.processTimeout()
+		}
+	}
+}
+
+func (l *ECSErrorLogger) Write(newError zapcore.Entry) {
+	if l.errorFile == nil {
+		if err := l.openExistingOrNewErrorFile(); err != nil {
+			log.Printf("[ecs error reporter] could not open error log file when write new error, err: %v", err)
+		}
+	}
+	newErrorLog := []byte(fmt.Sprintf("{%+v, Level:%+v, Caller:%+v, Message:%+v}\r\n",
+		newError.Time, newError.Level, newError.Caller, newError.Message))
+
+	if len(newErrorLog) > l.ErrorFileMaxSize {
+		log.Printf("[ecs error reporter] error size exceed the max size of error file")
+		return
+	}
+
+	err := l.errorFile.Truncate(0)
+	if err != nil {
+		log.Printf("[ecs error reporter] could not truncate error log file when write new error, err: %v", err)
+	}
+	l.size = 0
+
+	errorNum := len(l.queue)
+	for i := 0; i < errorNum; i++ {
+		e := l.queue[0]
+		if e.Message != newError.Message {
+			n, err := l.errorFile.Write([]byte(fmt.Sprintf("{%+v, Level:%+v, Caller:%+v, Message:%+v}\r\n",
+				e.Time, e.Level, e.Caller, e.Message)))
+			if err != nil {
+				log.Printf("[ecs error reporter] could not write error into error log file, err: %v", err)
+			}
+			l.size += n
+			l.queue = append(l.queue, e)
+		}
+		l.queue = l.queue[1:]
+	}
+
+	if l.size+len(newErrorLog) > l.ErrorFileMaxSize {
+		l.rotate(len(newErrorLog))
+	}
+
+	l.queue = append(l.queue, newError)
+	n, err := l.errorFile.Write(newErrorLog)
+	if err != nil {
+		log.Printf("[ecs error reporter] could not write new error into error log file, err: %v", err)
+	}
+	l.size += n
+}
+
+// rotate func could delete the old logs to make sure error log file less than 1 KB
+func (l *ECSErrorLogger) rotate(newErrorLogSize int) {
+	file, err := os.OpenFile(l.ErrorFilePath, os.O_CREATE|os.O_WRONLY|os.O_TRUNC|os.O_APPEND, 0777)
+	if err != nil {
+		log.Printf("[ecs error reporter] could not open error log file during rotation, err: %v", err)
```

Comment:
If you can't open the file and the function continues, `file.Truncate` would get a nil pointer .

Reply:
yeah, i just add return if open file met some error